### PR TITLE
Add `amp option` wp cli command to manage plugin options

### LINKED ISSUE #5783: Add WP-CLI commands for managing AMP settings
## Feature description

As demonstrated by #5782, there is a need for WP-CLI commands to manipulate the AMP settings, including:

* Setting the template mode.
* Setting the Reader theme.
* Toggling mobile redirection.
* Updating the supported post types (see https://github.com/ampproject/amp-wp/issues/5782#issuecomment-761877633).
* Toggling whether all templates are supported, and otherwise toggling individual templates.
* Updating suppressed plugins
* Setting the Paired URL structure (as of #5558) 

I'm specifically not including the manipulation of Analytics here, as we want to discourage it from being used.

In addition to updating the settings, we also need to be able to query the existing settings.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary

Fixes #5783 
Previously #7368

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: src/Cli/OptionCommand.php [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+				$option_name => $option_value,
+			]
+		);
+
+		if ( $response->as_error() ) {
+			/* translators: %s: option name */
+			WP_CLI::error( sprintf( __( 'Could not update option: %s', 'amp' ), $response->as_error()->get_error_message() ) );
+		}
+
+			/* translators: %s: option name */
```
Comment (westonruter) [inline src/Cli/OptionCommand.php]:
```suggestion
		/* translators: %s: option name */
```

File: src/Cli/OptionCommand.php [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+	 * @var ReaderThemes
+	 */
+	private $reader_themes;
+
+	/**
+	 * PluginSuppression instance.
+	 *
+	 * @var PluginSuppression
+	 */
+	private $plugin_suppression;
```
Comment (westonruter) [inline src/Cli/OptionCommand.php]: As noted in the other PR review, all references to plugin suppression can be removed.

File: src/Cli/OptionCommand.php [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+	 *
+	 * @subcommand list
+	 */
+	public function list_() {
+		$this->check_user();
+
+		$options = $this->get_options();
+
+		// Add reader themes to the options.
+		$options[ self::READER_THEMES ] = wp_list_pluck( $this->reader_themes->get_themes(), 'slug' );
```
Comment (westonruter) [inline src/Cli/OptionCommand.php]: See comment on other PR about moving this to another command.

**Review by westonruter — CHANGES_REQUESTED** [review]
Just pulling in this previous PR review: https://github.com/ampproject/amp-wp/pull/7368#pullrequestreview-1268132207

File: src/Cli/OptionCommand.php (hunk 2 above)
Comment (thelovekesh) [reply src/Cli/OptionCommand.php]: Aren't we going to add an option to update the suppressed plugins as per https://github.com/ampproject/amp-wp/issues/5783#issuecomment-1333577362?

File: src/Cli/OptionCommand.php (hunk 3 above)
Comment (thelovekesh) [reply src/Cli/OptionCommand.php]: Makes sense, and we can add it as a sub-command on the `amp option list` command as the` amp option list reader-themes`.

File: src/Cli/OptionCommand.php (hunk 2 above)
Comment (thelovekesh) [reply src/Cli/OptionCommand.php]: Skipping suppressed plugins to be managed by WP CLI for now.

**Comment by thelovekesh** [comment]
@westonruter For the PR, I included feature tests, but I'm not sure if I should also add unit tests. What do you think?

**Comment by westonruter** [comment]
I think feature tests are sufficient.

**Comment by thelovekesh** [comment]
@westonruter Just realized that helper function used in the command can have unit tests hence, added a few test cases in https://github.com/ampproject/amp-wp/pull/7430/commits/519152ce4b1de0cdab058d619869d013f2c86559

**Comment by westonruter** [comment]
Is this ready for re-review?

**Comment by thelovekesh** [comment]
> Is this ready for re-review?

@westonruter Yes.

File: .editorconfig [hunk 4]
```diff
@@ -11,9 +11,13 @@ trim_trailing_whitespace = true
...
 
-[{.babelrc,.eslintrc,.rtlcssrc,*.json,*.yml}]
+[*.{json,yml,feature}]
 indent_style = space
 indent_size = 2
 
 [*.md]
 trim_trailing_whitespace = false
+
+[{.babelrc,.eslintrc,.rtlcssrc}]
```
Comment (westonruter) [inline .editorconfig]: Curious why this is separated from the above block that has the same properties?

File: src/Cli/OptionCommand.php [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+
+	/**
+	 * Allowed options to be managed via the CLI.
+	 *
+	 * @var string[]
+	 */
+	const ALLOWED_OPTIONS = [
+		Option::READER_THEME,
+		Option::THEME_SUPPORT,
+		Option::MOBILE_REDIRECT,
```
Comment (westonruter) [inline src/Cli/OptionCommand.php:64]: I just realized that there is one other setting which would be quite helpful and that is whether sandboxing is enabled and the selected sandboxing level. Could you add those as well?

File: src/Cli/OptionCommand.php [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+		if ( $options instanceof WP_Error ) {
+			/* translators: %s: error message */
+			WP_CLI::error( sprintf( __( 'Could not retrieve options: %s', 'amp' ), $options->get_error_message() ) );
+		}
+
+		if ( ! isset( $options[ $option_name ] ) ) {
+			/* translators: %s: option name */
+			WP_CLI::error( sprintf( __( 'Could not update %s option. Does it exist?', 'amp' ), $option_name ) );
+		}
```
Comment (westonruter) [inline src/Cli/OptionCommand.php]: Why do we need to get all the options? This doesn't seem to be used.

File: src/Cli/OptionCommand.php [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+	 * - AMP Plugin options.
+	 * - AMP Plugin options that can be updated via the CLI.
+	 * - Reader themes.
+	 *
+	 * ## OPTIONS
+	 * [<cli-managed-options>]
+	 * : List of options that can be updated via the CLI.
+	 *
+	 * [<reader-themes>]
+	 * : List of Reader themes.
```
Comment (westonruter) [inline src/Cli/OptionCommand.php]: IMO, it may help with discoverability and simplify the code a bit if we make the reader theme list a sibling of the `list` command. So like this:

```
wp amp option list
wp amp option list-reader-themes
```

Then there won't be the need to check the `$args` and have custom command routing.

File: src/Cli/OptionCommand.php [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+				array_map(
+					static function ( $option_name, $option_value ) {
+						return compact( 'option_name', 'option_value' );
+					},
+					array_keys( $options ),
+					$options
+				),
+				[ 'option_name', 'option_value' ]
+			);
+		}
```
Comment (westonruter) [inline src/Cli/OptionCommand.php]: It may make sense to only every show the CLI-managed options, and then add a notice to list out the options that are managed via CLI.

File: src/Cli/OptionCommand.php [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+	 * Do a REST Request
+	 *
+	 * @param string $method HTTP method.
+	 * @param string $route REST route.
+	 * @param array  $assoc_args Associative args.
+	 *
+	 * @return \WP_REST_Response Response object.
+	 */
+	private function do_request( $method, $route, $assoc_args ) {
+		$request = new \WP_REST_Request( $method, $route );
```
Comment (westonruter) [inline src/Cli/OptionCommand.php]: Minor point. The `WP_REST_Request` could be imported above so we don't need to use `\` in this one place:

```suggestion
		$request = new WP_REST_Request( $method, $route );
```

File: src/Cli/OptionCommand.php (hunk 8 above)
Comment (westonruter) [reply src/Cli/OptionCommand.php]: If someone wants to get all the raw options, they can always just do:

```
wp option get amp-options
```

File: src/Cli/OptionCommand.php [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+	 * @var string
+	 */
+	const READER_THEMES = 'reader-themes';
+
+	/**
+	 * CLI managed options.
+	 *
+	 * @var string
+	 */
+	const CLI_MANAGED_OPTIONS = 'cli-managed-options';
```
Comment (westonruter) [inline src/Cli/OptionCommand.php]: Per my note below, I think we can just always only list the CLI-managed options and then include a note to say that other options aren't yet supported. There can be a link to GitHub for a user to request support for additional options.

**Review by westonruter — CHANGES_REQUESTED** [review]
Code looks good! Just a few more bits of feedback.

File: src/Cli/OptionCommand.php (hunk 5 above)
Comment (westonruter) [reply src/Cli/OptionCommand.php:64]: Oh, one more: Delete plugin data at uninstall

File: .editorconfig (hunk 4 above)
Comment (thelovekesh) [reply .editorconfig]: The last time I checked it was not working as expected with the same block but now it is.

I have reverted these changes. Also, I have removed the `.babelrc` and `.eslintrc` files since we are not using them anymore.

File: src/Cli/OptionCommand.php (hunk 6 above)
Comment (thelovekesh) [reply src/Cli/OptionCommand.php]: This is used to determine whether the user is trying to update the option which exists in `amp-options`. I added it before allowed options notice so that the user can learn that he is trying to update a non-existent option.

If we remove this guard then the user will see:
```php
if ( ! in_array( $option_name, self::ALLOWED_OPTIONS, true ) ) {
    /* translators: %s: option name */
    WP_CLI::error( sprintf( __( 'You are not allowed to update %s option via the CLI.', 'amp' ), $option_name ) );
}
```
for every attempt to update a wrong option.

Comment (thelovekesh) [reply src/Cli/OptionCommand.php]: I think we can update the message here to something like this:
```diff
if ( ! isset( $options[ $option_name ] ) ) {
    /* translators: %s: option name */
-   WP_CLI::error( sprintf( __( 'Could not update %s option. Does it exist?', 'amp' ), $option_name ) );
+   WP_CLI::error( sprintf( __( 'You are trying to update %s option, but it does not exist.', 'amp' ), $option_name ) );
}
```

Comment (westonruter) [reply src/Cli/OptionCommand.php]: I see. Yes, it can be removed now that we have `ALLOWED_OPTIONS`. And the update to the message looks good!

File: src/Cli/OptionCommand.php (hunk 8 above)
Comment (thelovekesh) [reply src/Cli/OptionCommand.php]: Updated the list command and it's more managed now 🎉 

image

File: src/Cli/OptionCommand.php (hunk 6 above)
Comment (thelovekesh) [reply src/Cli/OptionCommand.php]: IMO `ALLOWED_OPTIONS` notice will be a bit unclear when the user will try to update the wrong option i.e running `wp amp option update sandboxing_enable true` will result in `You are not allowed to update sandboxing_enable option via the CLI.` while with options check guard the message will `You are trying to update sandboxing_enable option, but it does not exist.`

Do you still want me to remove highlighted logic?

Comment (westonruter) [reply src/Cli/OptionCommand.php]: How about the message say something like:

```The option "%1$s" is not among the following options that can currently be managed via CLI: %2$s.```

Where the former is the provided option, and the latter is a comma-separated list of `ALLOWED_OPTIONS`

Comment (thelovekesh) [reply src/Cli/OptionCommand.php]: Perfect. Will update it 👍🏼

File: src/Cli/OptionCommand.php [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+
+		if ( ! in_array( $option_name, self::ALLOWED_OPTIONS, true ) ) {
+			/* translators: %1$s: option name, %2$s: list of allowed options */
+			WP_CLI::error( sprintf( __( 'The option "%1$s" is not among the following options that can currently be managed via CLI: %2$s', 'amp' ), $option_name, implode( ', ', self::ALLOWED_OPTIONS ) ) );
+		}
+
+		// Update type for some options.
+		if ( Option::SANDBOXING_LEVEL === $option_name ) {
+			$option_value = (int) $option_value;
+		}
```
Comment (westonruter) [inline src/Cli/OptionCommand.php:188]: Why was this specific cast needed? Is it because it is a string and the following logic requires it to be an int:

https://github.com/ampproject/amp-wp/blob/c28b8f1403116e831e837667cc7fca88e6fae9c3/src/Sandboxing.php#L123-L129

It seems like the `Sandboxing` service should be doing that cast to int as well, since it is already doing so for bool:

https://github.com/ampproject/amp-wp/blob/c28b8f1403116e831e837667cc7fca88e6fae9c3/src/Sandboxing.php#L120-L122

**Review by westonruter — APPROVED** [review]

File: src/Cli/OptionCommand.php (hunk 11 above)
Comment (thelovekesh) [reply src/Cli/OptionCommand.php:188]: I believe when we update options using REST API then it first validates the options(and return any validation failures) through provided schema. In the case of `'0' and '1'` and `'true' and 'false'` with schema type as boolean, it runs it through the `rest_sanitize_boolean` first hence type cast is not required but in the case of enums it fails:
https://github.com/ampproject/amp-wp/blob/c28b8f1403116e831e837667cc7fca88e6fae9c3/src/Sandboxing.php#L56
https://github.com/ampproject/amp-wp/blob/c28b8f1403116e831e837667cc7fca88e6fae9c3/src/Sandboxing.php#L68-L72


In other words, it doesn't reach here:
https://github.com/ampproject/amp-wp/blob/c28b8f1403116e831e837667cc7fca88e6fae9c3/includes/options/class-amp-options-manager.php#L368 

---

> It seems like the Sandboxing service should be doing that cast to int as well since it is already doing so for bool:

Type cast here is done by REST schema's `rest_sanitize_boolean` IMO.
